# poppr-poster-aps-2016

source data for the APS 2016 poster on Tools for analysis of clonal population genetic data in R

# Abstract

### [Tools for analysis of clonal population genetic data in R](http://www.apsnet.org/meetings/annual/abstracts/pages/abstractdetail.aspx?MID=816)

Knowledge of the population dynamics and evolution of plant pathogens allows
inferences on evolutionary processes involved in their adaptation to hosts,
pesticides, and other environmental pressures. These microbial pathogens
often require different tools for analysis due to the fact that they can be
clonal or partially clonal. With the advent of high-throughput sequencing
technologies, obtaining genome-wide data has become easier and cheaper than
ever before with techniques such as genotyping-by-sequencing. In 2013, we
created the widely used R package poppr for analysis of clonal populations.
In 2015, we published several additional extensions to poppr for use with
genomic data including the ability to define clones based on a genetic
distance threshold, minimum spanning networks with reticulation, and sliding
window analysis of the index of association. We present here an overview of
poppr as it pertains to traditional and high-throughput population genetic
data with select applications.
